polaris/polaris/sep10auth/views.py

```diff
@@ -89,20 +90,114 @@ def _transaction_from_post_request(request):
 
 
-def _validate_envelope_xdr(envelope_xdr):
+def _validate_challenge_xdr(envelope_xdr):
     """
     Validate the provided TransactionEnvelope XDR (base64 string). Return the
     appropriate error if it fails, else the empty string.
+
+    verify_challenge_transaction() does not verify that the transaction's
+    signers are valid and have the total weight greater or equal to the
+    source account's medium threshold value, so this function does that as
+    well.
     """
     try:
         verify_challenge_transaction(
             challenge_transaction=envelope_xdr,
             server_account_id=settings.SIGNING_KEY,
             network_passphrase=settings.STELLAR_NETWORK_PASSPHRASE,
         )
-    except InvalidSep10ChallengeError as err:
-        return str(err)
-    return ""
+    except InvalidSep10ChallengeError as e:
+        raise ValueError(str(e))
+
+    transaction_envelope = TransactionEnvelope.from_xdr(
+        envelope_xdr, settings.STELLAR_NETWORK_PASSPHRASE
+    )
+    # verify_challenge_transaction ensures transaction has at least
+    # one operation whose source is the client account.
+    source_account = transaction_envelope.transaction.operations[0].source
+    tx_hash = transaction_envelope.hash()
+
+    # make horizon /accounts API call to get signers and threshold
+    account_signers, threshold = _get_signers_and_threshold(source_account)
+
+    ##################################################################
+    # The functionality implemented below should go in stellar_sdk and
+    # invoked like so:
+    #
+    #       verify_challenge_transaction(
+    #           challenge_transaction=envelope_xdr,
+    #           server_account_id=settings.SIGNING_KEY,
+    #           network_passphrase=settings.STELLAR_NETWORK_PASSPHRASE,
+    #           account_signers=account_signers,
+    #           threshold=threshold
+    #       )
+    #
+    # Where account_signers and threshold are defined like so:
+    #
+    #       account_json = server.accounts().account_id(source_account).call()
+    #       account_signers = account_json["signers"]
+    #       threshold = account_json["thresholds"]["med_threshold"]
+    #
+    # This way, the client can choose which threshold the challenge should
+    # meet. It also makes sense to pass the unaltered account_json["signers"]
+    # instead of requiring the client to do some manipulation of the
+    # response before passing.
+    ##################################################################
+
+    # Get the account signers used for this transaction
+    server_kp = Keypair.from_public_key(settings.SIGNING_KEY)
+    matched_signers = []
+    for dec_signer in transaction_envelope.signatures:
+        try:
+            server_kp.verify(tx_hash, dec_signer.signature)
+        except BadSignatureError:
+            # dec_signer is not from the server's keypair
+            matched_signers.append(_match_signer(tx_hash, dec_signer, account_signers))
```

Comment:
❓ Will `_match_signer` append `None` to the list of `matched_signers` for any key that didn't match the server key, but did match one of the `account_signers`? Will that mean the resulting list of `matched_signers` will look something like this if there were a total of five signatures, three signers successfully matched and two were unrecognized:
```python
[
  {'keypair':'GD2WZVWK2K56Q4RWP5M2GTBHUIG56R7LWRUQMMZMK7I6TNLV72QDJEZ6', ...}, 
  {'keypair':'GD3AH24764GIOS2KUWHTIFOPRK75MB4ESLYU35JGXWKLEGKQWWEJCOJQ', ...},
  None,
  None,
  {'keypair':'GC33DJOHVMU6IVGZVDNWIMHUDP2LGARGSPBIBAV5FSESXZTPMHFHNS7V', ...}
]
```

Reply:
`_match_signer` would never return `None`. It checks if `matched_signer is None` and raises an error if it is.

Reply:
Oh, hmm. So if it encounters a signature that doesn't match any signer it fails immediately? Cool, sounds great!